packages/contracts-bedrock/src/L1/DelayedVetoable.sol

```diff
@@ -0,0 +1,159 @@
+    ///         inspection. We assume that nobody is able to trigger calls from address(0) during
+    ///         normal EVM execution.
+    modifier readOrHandle() {
+        if (msg.sender == address(0)) {
+            _;
+        } else {
+            // This WILL halt the call frame on completion.
+            _handleCall();
+        }
+    }
+
+    /// @custom:semver 0.0.1
+    /// @notice Sets the target admin during contract deployment.
+    /// @param vetoer_ Address of the vetoer.
+    /// @param initiator_ Address of the initiator.
+    /// @param target_ Address of the target.
+    /// @param delay_ Address of the delay.
+    constructor(address vetoer_, address initiator_, address target_, uint256 delay_) Semver(0, 0, 1) {
+        _vetoer = vetoer_;
+        _initiator = initiator_;
+        _target = target_;
+        _delay = delay_;
+    }
+
+    /// @notice Gets the initiator
+    /// @return Initiator address.
+    function initiator() external readOrHandle returns (address) {
+        return _initiator;
+    }
+
+    //// @notice Queries the vetoer address.
+    /// @return Vetoer address.
+    function vetoer() external readOrHandle returns (address) {
+        return _vetoer;
+    }
+
+    //// @notice Queries the target address.
+    /// @return Target address.
+    function target() external readOrHandle returns (address) {
+        return _target;
+    }
+
+    /// @notice Gets the delay
+    /// @return Delay address.
+    function delay() external readOrHandle returns (uint256) {
+        return _delay;
+    }
+
+    // TODO(maurelian): Remove this? The contract currently cannot handle forwarding ETH and I'm
+    //   not sure the complexity is warranted.
+    //   If we do allow it:
+    //      1. the callHash will need to include the value
+    //      2. forwarding will need to be done by passing the callHash, rather than the unhashed data
+    /// @notice Used when no data is passed to the contract.
+    receive() external payable {
+        _handleCall();
+    }
+
+    /// @notice Used for all calls that pass data to the contract.
+    fallback() external payable {
+        _handleCall();
+    }
+
+    /// @notice Receives all calls other than those made by the vetoer.
+    ///         This enables transparent initiation and forwarding of calls to the target and avoids
+    ///         the need for additional layers of abi encoding.
+    function _handleCall() internal {
+        bytes32 callHash = keccak256(msg.data);
+
+        // Case 1: The initiator is calling the contract to initiate a call.
+        if (msg.sender == _initiator && _queuedAt[callHash] == 0) {
+            _queuedAt[callHash] = block.timestamp;
+            emit Initiated(callHash, msg.data);
+            return;
+        }
+
+        // Case 2: The vetoer is calling the contract to veto a call.
+        if (msg.sender == _vetoer && _queuedAt[callHash] != 0 && block.timestamp <= _queuedAt[callHash] + _delay) {
```

Comment:
The vetoer has the ability to lose its ability to veto on maturation?

Reply:
yeah, I kind of think that's the right approach. 
IMO this basically just makes the correct behavior (of the vetoer) explicit.

Reply:
Hmm.... For short delay periods this does make censorship attacks against the vetoer a bit easier.
I now lean towards removing this to allow for a bidding war to persist past the delay, wdyt?

Reply:
I ended up making that change and including a comment to clarify the rationale.